File: Segundo_Curso/Segundo_Cuatri/PDOO/java/src/irrgarten/LabyrinthCharacter.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 *
 * @author el_dramas
 */
public abstract class LabyrinthCharacter {
    
    static private final int INVALID_POS = -1;
    
    private String name;
    private float intelligence;
    private float strength;
    private float health;
    private int row;
    private int col;
    
    public LabyrinthCharacter(String name,float intelligence,float strength, float health){
        this.name = name;
        this.intelligence = intelligence;
        this.strength = strength;
        this.health = health;
        
        //TODO: Preguntar si es así.
        this.row = INVALID_POS;
        this.col = INVALID_POS; 
    }
    
    public LabyrinthCharacter(LabyrinthCharacter other){
        
    }
    
    boolean dead(){
        return this.health <= 0;
    }
    
    public int getRow(){
        return this.row;
    }
    
    public int getCol(){
        return this.col;
    }
    
    protected float getIntelligence(){
        return this.intelligence;
    }
    
    protected float getStrength(){
        return this.strength;
    }
    
    protected float getHealth(){
        return this.health;
    }
    
    protected void setHealth(float health){
        this.health = health;
    }
    
    public void setPos(int row, int col){
        this.row = row;
        this.col = col;
    }
    
    public String toString(){
        return "[" + name + ", " + health +" HP," + strength + " SP, " +  intelligence + " IP, " + "(" + row + ", " +col + ")]";
    }
    
    protected void gotWounded(){
        this.health--;
    }
    
    public abstract float attack();
    
    public abstract boolean defend(float attack);
}

File: Segundo_Curso/Segundo_Cuatri/PDOO/ruby/Irrgarten/LabyrinthCharacter.rb
module Irrgarten
  class LabyrinthCharacter
    @@INVALID_POS = -1

    # def initialize(name,intelligence,strength,health)
    #   @name = name
    #   @intelligence = intelligence
    #   @strength = strength
    #   @health = health
    #   @row = @@INVALID_POS
    #   @col = @@INVALID_POS
    # end

    #Constructor privado porque se pretende que sea abstracta / no instanciable
    private_class_method :new

    attr_reader :name
    attr_reader :row
    attr_reader :col

    def self.newCopy(other)
      new(other.name,other.intelligence, other.strength, other.health)
      @row = other.row
      @col = other.col
    end

    def dead
      @health<=0
    end

    protected
      def intelligence
        @intelligence
      end

      def strength
        @strength
      end

      def health
        @health
      end

      def health=(health)
        @health = health
      end

      def gotWounded
        @health -= 1
      end
    public
      def pos=(row,col)
        if (row >= 0) && (col >= 0) then
          @row = row
          @col = col
        end
      end

      def to_s
        "[#{@name}, #{@health} HP, #{@intelligence} IP, #{@strength} SP,(#{@row},#{@col})]\n"
      end

      #TODO?: Como se hacen metodos abstractos en ruby
      def attack
      end

      def defend
      end

  end#class
  lc = LabyrinthCharacter.new("ª",1,5,9)
  lc.to_s
  lc_1 = LabyrinthCharacter.newCopy(lc)
  lc_1.to_s
end#module
